src/modules/ModuleRunner.ts

```diff
@@ -17,31 +17,142 @@ limitations under the License.
 import { safeSet } from "matrix-js-sdk/src/utils";
 import { TranslationStringsObject } from "@matrix-org/react-sdk-module-api/lib/types/translations";
 import { AnyLifecycle } from "@matrix-org/react-sdk-module-api/lib/lifecycles/types";
+import {
+    DefaultCryptoSetupExtensions,
+    ProvideCryptoSetupExtensions,
+} from "@matrix-org/react-sdk-module-api/lib/lifecycles/CryptoSetupExtensions";
+import {
+    DefaultExperimentalExtensions,
+    ProvideExperimentalExtensions,
+} from "@matrix-org/react-sdk-module-api/lib/lifecycles/ExperimentalExtensions";
 
 import { AppModule } from "./AppModule";
 import { ModuleFactory } from "./ModuleFactory";
 
 import "./ModuleComponents";
 
+class ExtensionImplementationMap {
+    public hasDefaultCryptoSetupExtension: boolean = true;
+    public hasDefaultExperimentalExtension: boolean = true;
+}
+
+/**
+ * Handles and manages any extensions provided by modules
+ */
+class ExtensionsManager {
+    private _cryptoSetup: ProvideCryptoSetupExtensions;
+    private _experimental: ProvideExperimentalExtensions;
+    private _implementionMap: ExtensionImplementationMap = new ExtensionImplementationMap();
+
+    /**
+     * Create a new instance
+     */
+    public constructor() {
+        // Set up defaults
+        this._cryptoSetup = new DefaultCryptoSetupExtensions();
+        this._experimental = new DefaultExperimentalExtensions();
+    }
+
+    /**
+     * Provides a crypto setup extension.
+     *
+     * @returns The registered extension. If no module provides this extension, a default implementation is returned
+     */
+    public get cryptoSetup(): ProvideCryptoSetupExtensions {
+        return this._cryptoSetup;
+    }
+
+    /**
+     * Provides a n experimental extension.
+     *
+     * @remarks
+     * This method extension is provided to simplify experimentaion an development, and is not intended for production code
+     *
+     * @returns The registered extension. If no module provides this extension, a default implementation is returned
+     */
+    public get experimental(): ProvideExperimentalExtensions {
+        return this._experimental;
+    }
+
+    /**
+     * Resets the extension to the defaults
+     *
+     * Intended for test usage only.
+     */
+    public reset(): void {
+        this._implementionMap = new ExtensionImplementationMap();
+        this._cryptoSetup = new DefaultCryptoSetupExtensions();
+        this._experimental = new DefaultExperimentalExtensions();
+    }
+
+    /**
+     * Add any extensions provided by the module
+     *
+     * @param module - The appModule to check for extensions
+     *
+     * @throws if an extension is provided by more than one module
+     *
+     */
+    public addExtensions(module: AppModule): void {
+        const runtimeModule = module.module;
+
+        /* Record the cryptoSetup extensions if any */
```

Comment:
We seem to be inconsistent about whether  `_cryptoSetup` is a single extension (for example, `get cryptoSetup` describes it as "a crypto setup extension") or several ("Record the cryptoSetup extension**s**"). 

Since we enforce that there is only one instance, it seems logical to me to consider it a *single* extension.

Reply:
You are correct. Each module can provide multiple extensions, but only one of each type. 

I will change the wording.

Reply:
A monorepo which includes matrix-react-sdk-module-api would be handy here, to make changes to module-api and react-sdk in the same committ. But I guess that opens a whole new can of worms?

Reply:
> Not sure how to fix this securely though, as both matrix-react-sdk and matrix-react-sdk-module-api will need to change.

Yes, it's hard to do, and probably not worthwhile. Let's just be consistent for *new* code.

> A monorepo which includes matrix-react-sdk-module-api would be handy here, to make changes to module-api and react-sdk in the same committ. But I guess that opens a whole new can of worms?

It does a bit, yes. In any case, we need to be very careful about making breaking changes to `matrix-react-sdk-module-api`: after all, its whole point is to provide a stable API that modules can rely on. Generally, I'd say this sort of scenario won't arise very often.